surfacers/cloudwatch/cloudwatch.go

```diff
@@ -0,0 +1,240 @@
+			}
+
+		case *metrics.Distribution:
+			for i, distributionBounds := range value.Data().LowerBounds {
+				dimensions := append(emLabelsToDimensions(em), &cloudwatch.Dimension{
+					Name:  aws.String(distributionDimensionName),
+					Value: aws.String(strconv.FormatFloat(distributionBounds, 'f', -1, 64)),
+				})
+
+				cw.publishMetrics(cw.newCWMetricDatum(metricKey, float64(value.Data().BucketCounts[i]), dimensions, em.Timestamp))
+			}
+
+		default:
+			continue LoopEventMetrics
+		}
+
+	}
+}
+
+// publish the metrics to cloudwatch, using the namespace provided from configuration
+func (cw *CWSurfacer) publishMetrics(md *cloudwatch.MetricDatum) {
+	if len(cw.cwMetricDatumCache) >= 20 {
+		_, err := cw.session.PutMetricData(&cloudwatch.PutMetricDataInput{
+			Namespace:  aws.String(cw.c.GetNamespace()),
+			MetricData: cw.cwMetricDatumCache,
+		})
+
+		if err != nil {
+			cw.l.Errorf("Failed to publish metrics to cloudwatch: %s", err)
+		}
+
+		cw.cwMetricDatumCache = cw.cwMetricDatumCache[:0]
+	}
+
+	cw.cwMetricDatumCache = append(cw.cwMetricDatumCache, md)
+}
+
+// calculateFailureMetrics calculates a failure cumalative metric, from the
+// total and success metrics in the eventmetric.
+func calculateFailureMetric(em *metrics.EventMetrics) (float64, error) {
+	successMetric, totalMetric := em.Metric("success"), em.Metric("total")
+
+	success, successOK := successMetric.(metrics.NumValue)
+	total, totalOK := totalMetric.(metrics.NumValue)
+
+	if !successOK || !totalOK {
+		return 0, fmt.Errorf("unexpected error, either success or total is not a number")
+	}
+
+	failure := total.Float64() - success.Float64()
+
+	return failure, nil
+}
+
+// determine if we should ignore the prober type label, based on the ignoreLabelsRegex config
+func (cw *CWSurfacer) ignoreProberTypeLabel(em *metrics.EventMetrics) bool {
+	if cw.ignoreLabelsRegex != nil {
+		if cw.ignoreLabelsRegex.MatchString(em.Label("ptype")) {
+			return true
+		}
+	}
+
+	return false
+}
+
+// Create a new cloudwatch metriddatum using the values passed in.
+func (cw *CWSurfacer) newCWMetricDatum(metricname string, value float64, dimensions []*cloudwatch.Dimension, timestamp time.Time) *cloudwatch.MetricDatum {
+	metricDatum := cloudwatch.MetricDatum{
+		Dimensions:        dimensions,
+		MetricName:        aws.String(metricname),
+		Value:             aws.Float64(value),
+		StorageResolution: aws.Int64(cw.c.GetResolution()),
+		Timestamp:         aws.Time(timestamp),
+	}
+
+	unitTypes := map[string]string{
+		"latency":  cloudwatch.StandardUnitMilliseconds,
+		"failures": cloudwatch.StandardUnitCount,
+		"success":  cloudwatch.StandardUnitCount,
+		"total":    cloudwatch.StandardUnitCount,
+		"timeouts": cloudwatch.StandardUnitCount,
+	}
+
+	if value, exists := unitTypes[metricname]; exists {
+		metricDatum.Unit = aws.String(value)
+	}
+
+	// distributions are of unit type count, identify if the dimensions contain a distribution key
```

Comment:
I am curious why we need to enforce the count unit. You're anyway converting distribution to regular metrics. I think  it's safe assume StandardUnitCount for all metrics except for latency, similar to how stackdriver handles it.

Reply:
We don't, but I honestly wasn't sure on the best way forward here - that was my first approach, but then I thought it's better to be explicit. Let me rework this.

Reply:
Ok reworked logic here: https://github.com/google/cloudprober/pull/583/commits/66083406b9a384b9a98fb876b74a157a0bd1acda

Reply:
Going to resolve this conversation, please re-open if there's something I've missed.